### DISCUSSION TITLE
STARTTLS Support

### DISCUSSION THREAD

**Discussion by kaaax0815** [BetaHuhn/ackee-report#44] [Ideas] [answered]
How can I send the Email over Port 587 (STARTTLS) instead of default 487 (TLS)?
Using `tls://` doesnt work :(

**Comment by BetaHuhn** [comment]
Hey!

You can specify your email servers port on the first run of [ackee-report](https://github.com/BetaHuhn/ackee-report). The program will ask you to input a few options (like your Ackee host), including the port of your email server. 

Just enter the port i.e. 587, when asked to input the "email port".  You can also change the port later in the config file (run `ackee-report config` to see where the file is stored).

This is also described in the [documentation](https://github.com/BetaHuhn/ackee-report#%EF%B8%8F-configuration) under [email-setup](https://github.com/BetaHuhn/ackee-report#email-setup).

Does this answer your question?

**Comment by kaaax0815** [comment, marked as answer]
Config (redacted)
```json
{
	"ackee": {
		"token": "68d9341f-1ab2-4514-8cb8-f0b138d34573",
		"server": "https://ackee.server.kaaaxcreators.de/"
	},
	"email": {
		"port": "587",
	}
}
```
Command: `ackee-report email -d all -t <email>`
```
- Generating email...(node:2428) UnhandledPromiseRejectionWarning: Error: 1948:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:c:\ws\deps\openssl\openssl\ssl\record\ssl3_record.c:332:

(node:2428) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:2428) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
Turning on smtps in postfix fixed it, but still a feature request

**Reply by BetaHuhn** [reply]
Under the hood [ackee-report](/) uses [Nodemailer](https://nodemailer.com/smtp/) which should automatically upgrade to STARTTLS if the server supports it.

What option from [Nodemailer](https://nodemailer.com/smtp/) could we expose to fix your issue?

**Reply by kaaax0815** [reply]
In the doc
the use `secure: false, // upgrade later with STARTTLS` for 587 and `secure: true, // use TLS` for 465

**Reply by kaaax0815** [reply]
I tried to try it myself with `node src/index.js -d all -t <email>`, but i get `error: unknown option '-d'`. On windows 10

**Reply by kaaax0815** [reply]
Even when installing it globally i get the same error @BetaHuhn

**Reply by kaaax0815** [reply]
Kann eigentlich auch Deutsch schreiben lul, aber das es andere verstehen. I used the latest release and using `secure: false, // upgrade later with STARTTLS` starttls works fine

**Reply by BetaHuhn** [reply]
> In the doc they use secure: false, // upgrade later with STARTTLS for 587 and secure: true, // use TLS for 465

I think the best approach would be to set `secure: false` automatically if the port is 587 or 25 and else set it to true.

But as it is false by default it shouldn't make a difference in your case.

> I tried to try it myself with node src/index.js -d all -t <email>, but i get error: unknown option '-d'. On windows 10

What were you trying to achieve? You have to run the program like this:

```
ackee-report email -d all -t <email>
```

**Reply by kaaax0815** [reply]
i runned the program differently because i cloned the github. yes you are right because smtps breaks when using `secure: false`

**Reply by BetaHuhn** [reply]
Ah okay. So it works if you manually set secure to true when using port 587?

**Reply by kaaax0815** [reply]
secure to false on port 587 xd then starttls works but smtps not anymore, so a new parameter, protocol detection or port detection would do it

**Reply by kaaax0815** [reply]
a simple if statement would be working, `if port == 25 || port == 587` because port 25 is also smtp. and switching between unsecure and starttls on port 25 does nodemailer itself

**Reply by BetaHuhn** [reply]
So if I'm understanding it correctly, if the port is 25 or 587 we should set secure to false so Nodemailer will use STARTTLS, and for any other port we should set it to true?

**Reply by kaaax0815** [reply]
yes

**Reply by BetaHuhn** [reply]
Shouldn't any email server which supports TLS directly also support STARTTLS?

Since your initial issue was to use STARTTLS, which ackee-report currently seems to do because secure is false by default, I'm just wondering if the change is really necessary

I haven't had any issues when using port 465 with the current version i.e. secure being set to false.

**Reply by kaaax0815** [reply]
As of Master secure is set to true. And starttls and tls are email wise two different protocols. Using postfix smtps you can't connect to it using starttls

**Reply by kaaax0815** [reply]
You can read this https://kinsta.com/de/blog/smtp-port-wahlen/. I have to correct myself. Port 465 uses ssl not tls https://www.ionos.de/digitalguide/server/sicherheit/tls-vs-ssl/

**Reply by BetaHuhn** [reply]
https://github.com/BetaHuhn/ackee-report/blob/01ab2e37d64ed948c0582d5f6b39c03cf1cfff04/src/service/email.js#L9-L15

Ah, now I understand the issue. I didn't realize that the secure option is explicitly set to true in ackee-reports code.

I find this sentence in Nodemailers docs rather confusing: "if true the connection will use TLS when connecting to server. If false (the default) then TLS is used if server supports the STARTTLS extension."

When secure is set to true, Nodemailer will initiate the connection to the server over a secured connection, and with secure false it will connect unsecured and then upgrade if the servers supports it with STARTTLS, right?

What I understand from those articles is, that secure only needs to be turned on for port 465 as it uses SSL/TLS directly, and all other ports should support STARTTLS and secure should be turned off? If that's the case, a simple if statement, as you mentioned, we be enough.

Am I finally understanding it correctly? :D

**Reply by kaaax0815** [reply]
I think I and you are right. But ofc I don't can't promise it's right. But I would say that you are right, because i tested it with my email server

**Reply by BetaHuhn** [reply]
Okay, nice teamwork then 😄

Would you be willing to implement this change yourself?

**Reply by kaaax0815** [reply]
Do you mean create this feature and create a pull request?

**Reply by BetaHuhn** [reply]
Create a pull request in which you implement the change/feature.
